### DISCUSSION TITLE
ScriptReference/Mathf.Lerp

### DISCUSSION THREAD

**Discussion by giscus** [Developer-Notes-Extension/Unity-Notes#557] [Notes]
# ScriptReference/Mathf.Lerp

https://docs.unity3d.com/ScriptReference/Mathf.Lerp.html

**Comment by vertxxyz** [comment]
If you're looking for easing functions to transform your `t` value [`UnityEngine.UIElements.Experimental.Easing`](https://docs.unity.cn/ScriptReference/UIElements.Experimental.Easing.html) exists.
If you're not comfortable using an experimental API, you can find easing functions online, or you can copy the [source](https://github.com/Unity-Technologies/UnityCsReference/blob/e740821767d2290238ea7954457333f06e952bad/ModuleOverrides/com.unity.ui/Core/EasingCurves.cs) (keep it in your project to avoid license issues) to maintain it yourself.

**Comment by FreyaHolmer** [comment]
##### Lerp Smoothing
If you do lerp-smoothing (dynamic exponential ease-out with a moving target) like this:
```cs
void Update(){
    value = Mathf.Lerp( value, target, factor ); // this is framerate dependent :(
}
```
*The rate of decay will be different for different framerates*, which is usually not what you want. Here are some framerate independent variants

##### Method 1 (express in terms of half-life)
```cs
value = Mathf.Lerp( value, target, 1-Mathf.Pow(2,-Time.deltaTime/h) );
```
`h` is the half-life $\lambda$, the time you want it to take to nominally get halfway to the target.

##### Method 2 (express in terms of remaining fraction per second)
```cs
value = Mathf.Lerp( value, target, 1-Mathf.Pow( c, Time.deltaTime ) );
```
$c$ is the remaining fraction of distance per second. If $c = 0.25$, then after 1 second, it will have 25% of the distance remaining

##### Conversion/Porting
Given your existing framerate-dependent setup where
$R$ is the current/ideal framerate where the old smoothing looked nice, such as $R=60$ for 60 fps
$f$ is the old smoothing factor you used in `a = lerp( a, b, f )`

Then,
For method 1: $\lambda = -1/(R\cdot\log_2(1-f))$ [(calculator)](https://www.wolframalpha.com/input?i=lambda+%3D++-1%2F%28R*log2%281-f%29%29%2C+where+f+%3D+0.05%2C+R+%3D+60)
For method 2: $c = (1-f)^R$ [(calculator)](https://www.wolframalpha.com/input?i=c+%3D++%281-f%29%5ER%2C+where+f+%3D+0.05%2C+R+%3D+60)

Note that both $c$ and $\lambda$ are constants, so avoid calculating them every frame!

**Comment by QuinnCG** [comment]
Some alternative techniques, for the purposes of smoothing, are: [`Mathf.SmoothDamp`](https://docs.unity3d.com/ScriptReference/Mathf.SmoothDamp.html), [`Vector2.SmoothDamp`](https://docs.unity3d.com/ScriptReference/Vector2.SmoothDamp.html), and [`Vector3.SmoothDamp`](https://docs.unity3d.com/ScriptReference/Vector3.SmoothDamp.html).
```csharp
Vector3 velocity;

void Update()
{
	var target = new Vector3(0f, 5f, 0f);
	float duration = 0.2f;
	transform.position = Vector3.SmoothDamp(transform.position, target, ref velocity, duration);
	// Note: that "duration" is the number of seconds to go from the start to the target position.
}
```
